src/main/java/com/iexec/core/task/update/TaskUpdateRequestManager.java

```diff
@@ -89,35 +109,61 @@ public void setRequestConsumer(final TaskUpdateRequestConsumer consumer) {
         if (consumer == null){
             log.warn("Waiting for consumer before consuming [queueSize:{}]", queue.size());
             return;
         }
 
-        while (!Thread.currentThread().isInterrupted()){
-            waitForTaskUpdateRequest();
+        // We should only start threads to complete the thread pool.
+        // This is useless for a fresh start but could be handy
+        // if current thread is stopped and current method is rescheduled.
+        final int nbThreadsToStart = TASK_UPDATE_THREADS_POOL_SIZE - updatingThreads.get();
+        for (int i = 0; i < nbThreadsToStart; i++) {
+            startTaskUpdateThread();
         }
     }
 
+    private void startTaskUpdateThread() {
+        // When a task update is achieved or if there's an issue,
+        // we can start another task update.
+        // This should ensure a constant number of waiting and running threads.
+        CompletableFuture.runAsync(this::waitForTaskUpdateRequest, taskUpdateExecutor)
+                .thenRunAsync(this::startTaskUpdateThread);
+    }
+
     /**
      * Wait for new task update request and run the update once a request is arrived.
      * 2 requests can be run in parallel if they don't target the same task.
      * <br>
      * Interrupts current thread if queue taking has been stopped during the execution.
      */
     private void waitForTaskUpdateRequest() {
+        updatingThreads.incrementAndGet();
         log.info("Waiting requests from publisher [queueSize:{}]", queue.size());
         try {
-            String chainTaskId = queue.take();
-            CompletableFuture.runAsync(() -> {
-                synchronized (locks.computeIfAbsent(chainTaskId, key -> new Object())){ // require one update on a same task at a time
-                    consumer.onTaskUpdateRequest(chainTaskId); // synchronously update task
-                }
-            }, taskUpdateExecutor);
+            final Task task = queue.take();
+            String chainTaskId = task.getChainTaskId();
+            log.info("Selected task [chainTaskId: {}, status: {}]", chainTaskId, task.getCurrentStatus());
+            synchronized (locks.computeIfAbsent(chainTaskId, key -> new Object())) { // require one update on a same task at a time
+                consumer.onTaskUpdateRequest(chainTaskId); // synchronously update task
+            }
         } catch (InterruptedException e) {
             log.error("The unexpected happened", e);
             Thread.currentThread().interrupt();
+        } finally {
+            updatingThreads.decrementAndGet();
         }
     }
+
+    PriorityBlockingQueue<Task> createQueue() {
+        // Tasks whose status are the more advanced should be computed before others
+        final Comparator<Task> comparator = Comparator.comparing((Task task) -> task.getCurrentStatus().ordinal(), Comparator.reverseOrder())
+                .thenComparing(Task::getContributionDeadline, Comparator.reverseOrder());
```

Comment:
Shouldn't it be `Comparator.naturalOrder()` for **contributionDeadline** ?
I mean we want to prioritize tasks with a short contribution deadline and need to check we have the correct behaviour.
Second question, do we prioritize by status, then deadline or the other way around ?

Reply:
You're absolutely right! The condition has been reversed. Thanks!

I think we should prioritize by status then deadline because there could be an attack otherwise: a big task could be never transitioned if lots of small tasks are asked. Then, the big task would fall into contribution timeout, although it has been computed by one or more workers.

Reply:
Thanks, I get the idea !
After reading https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Enum.html#compareTo(E),
I think the first comparison could be written `Comparator.comparing(Task::getCurrentStatus, Comparator.reverseOrder())`
There is no requirement to explicitly call `ordinal()` as it is handled through the `compareTo()` method.
See also: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Comparable.html

Reply:
Absolutely! Fixed in https://github.com/iExecBlockchainComputing/iexec-core/pull/479/commits/1cd00c4723e626da108e1586e1814a2f0fca79ee :)